app/src/ui/diff/text-diff.tsx

```diff
@@ -480,17 +499,138 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
+      // We can only infer which line was clicked when the context menu is opened
+      // via a mouse event.
+      return null
+    }
+
+    if (!this.props.onDiscardChanges) {
+      return null
+    }
+
+    const lineNumber = editor.lineAtHeight(event.y)
+    const diffLine = diffLineForIndex(this.props.diff.hunks, lineNumber)
+    if (diffLine === null || !diffLine.isIncludeableLine()) {
+      // Do not show the discard options for lines that are not additions/deletions.
+      return null
+    }
+
+    const range = findInteractiveDiffRange(this.props.diff.hunks, lineNumber)
+    if (range === null) {
+      return null
+    }
+    const hunkType = this.getHunkType(range)
+    if (hunkType === null) {
+      return null
+    }
+
+    // When user opens the context menu from the hunk handle, we should
+    // discard the range of changes that from that hunk.
+    if (targetHasClass(event.target, 'hunk-handle')) {
+      return [
+        {
+          label: this.getDiscardLabel(hunkType, range.to - range.from + 1),
+          action: () => this.onDiscardChanges(file, range.from, range.to),
+        },
+      ]
+    }
+
+    // When user opens the context menu from a specific line, we should
+    // discard only that line.
+    if (targetHasClass(event.target, 'diff-line-number')) {
+      // We don't allow discarding individual lines on hunks that have both
+      // added and modified lines, since that can lead to unexpected results
+      // (e.g discarding the added line on a hunk that is a 1-line modification
+      // will leave the line deleted).
+      return [
+        {
+          label: this.getDiscardLabel(hunkType, 1),
+          action: () => this.onDiscardChanges(file, lineNumber),
+          enabled: hunkType !== HunkType.Modified,
+        },
+      ]
+    }
+
+    return null
+  }
+
+  private getHunkType(range: { from: number; to: number }): HunkType | null {
+    const startDiffLine = diffLineForIndex(this.props.diff.hunks, range.from)
+    const endDiffLine = diffLineForIndex(this.props.diff.hunks, range.to)
+
+    if (startDiffLine === null || endDiffLine === null) {
+      return null
+    }
+
+    if (startDiffLine.type !== endDiffLine.type) {
+      return HunkType.Modified
+    }
+    if (startDiffLine.type === DiffLineType.Add) {
+      return HunkType.Added
+    }
+    if (startDiffLine.type === DiffLineType.Delete) {
+      return HunkType.Deleted
+    }
+
+    return null
+  }
+
+  private onDiscardChanges(
+    file: WorkingDirectoryFileChange,
+    startLine: number,
+    endLine: number = startLine
+  ) {
+    if (!this.props.onDiscardChanges) {
+      return
+    }
+
+    const selection = file.selection
+      .withSelectNone()
+      .withRangeSelection(startLine, endLine - startLine + 1, true)
+
+    this.props.onDiscardChanges(this.props.diff, selection)
+  }
+
+  private getDiscardLabel(hunkType: HunkType, numLines: number): string {
+    return toPlatformCase(
```

Comment:
I'd rather we not use `toPlatformCase`. It has a checkered past in Desktop and I believe the last time we discussed it we were all in favor of removing it so adding more usage to it would be counterproductive.

For context the rules for title casing are much too complex to succinctly express in a helper function like this which is why you see these ternary statements scattered about the code base. See https://github.com/desktop/desktop/pull/5841#pullrequestreview-162478757 and https://github.com/desktop/desktop/issues/5847 for some of the back story.

In this scenario it's possible that it's doing the right thing but while I'm far far far from an expert in English grammar I wonder if the word "these" can be considered an article in this sentence and therefore should be excluded from uppercasing based on https://developer.apple.com/library/archive/documentation/FinalCutProX/Conceptual/FxPlugHIG/TextStyleGuidelines/TextStyleGuidelines.html

> In title-style capitalization, capitalize every word except:
> - Articles (a, an, the)
>

I propose we make this easy on us by rewording and refactor here 👇 

```ts
private getDiscardLabel(hunkType: HunkType, numLines: number): string {
  const s = numLines > 1 ? 's' : ''
  return __DARWIN__ ? `Discard Line${s}` : `Discard line${s}`
}
```

or, slightly more verbose

```ts
private getDiscardLabel(hunkType: HunkType, numLines: number): string {
  if (numLines > 1) {
    return __DARWIN__ ? 'Discard Lines' : 'Discard lines'
  } else {
    return __DARWIN__ ? 'Discard Line' : 'Discard line'
  }
}
```

Reply:
Good point! I just realized that this method was not used anywhere else in the app 🤦‍♂️ (don't know how I even discovered it).

If you don't mind I'll send another PR to completely remove that method from the codebase.

Reply:
> If you don't mind I'll send another PR to completely remove that method from the codebase.

I don't mind, quite the opposite 😄